### ISSUE TITLE
Feature request: Read only Singularity cache

### ISSUE THREAD

**Issue by ewels** [nextflow-io/nextflow#1879]
## New feature

Suggestion is to add a new Nextflow environment variable which works alongside `$NXF_SINGULARITY_CACHEDIR` with one difference: it is assumed to be read-only. It could be called `$NXF_SINGULARITY_LIBRARY` or something.

## Usage scenario 

With DSL2 we are relying on `$NXF_SINGULARITY_CACHEDIR` over at @nf-core much more, as users can no longer run pipelines using `-with-singularity`. Generally it works great.

On large shared multi-user HPC setups it makes sense to have a shared singularity cachedir for all users, to avoid duplicating many GBs of identical containers (it also makes it easier and faster for users to run pipelines). It's not a good idea to have this world writeable in such situations though. Since https://github.com/nextflow-io/nextflow/issues/1706 was closed, read-only cache dirs do work and the vision of a shared cachedir is possible 🤩  ('cc @pontus).

The problem is when a user tries to run a pipeline that _hasn't_ already been cached. Then Nextflow will try to pull the required containers and save them to `$NXF_SINGULARITY_CACHEDIR` and fail if it's read-only. My suggestion is to add a new env var that allows the user to keep their own personal `$NXF_SINGULARITY_CACHEDIR` which they can write to _and_ also make use of a centralised shared read-only singularity cache.

## Suggest implementation 

I guess the code doesn't need to be super complicated, just a case of looping through more than one `localpath` here:

https://github.com/nextflow-io/nextflow/blob/671ae6d85df44f906747c16f6d73208dbc402d49/modules/nextflow/src/main/groovy/nextflow/container/SingularityCache.groovy#L177-L183

Looking at the code we could probably avoid needing to add additional env variables if you prefer if _all three_ of these were checked for existing images instead of just the single top priority one:

https://github.com/nextflow-io/nextflow/blob/671ae6d85df44f906747c16f6d73208dbc402d49/modules/nextflow/src/main/groovy/nextflow/container/SingularityCache.groovy#L115-L118

Open to whatever solution you suggest.

**Comment by pditommaso** [comment]
It sounds double, only concern is trying to avoid the proliferation of magic env variables. What about adding a option in the config file eg. `singularity.cacheReadOnly`. ?

**Comment by ewels** [comment]
Understandable 👍🏻  I'd _prefer_ an env variable if possible though.. In our use case I'm thinking of having @nf-core pipelines set up as environment modules on the cluster, so that people can do `module load nf-core-rnaseq` on our secure offline and have everything that they need. An env variable is easy to do with this, a nextflow config option is maybe doable but would require workarounds and hacks.

**Comment by pditommaso** [comment]
Then we can have a compromise and have both 😄 

It would be enough adding a similar method `getLibraryDir` 

https://github.com/nextflow-io/nextflow/blob/dd92772bc71d0b01c3821bd24197d40a05a9e69b/modules/nextflow/src/main/groovy/nextflow/container/SingularityCache.groovy#L124-L151

**Comment by ewels** [comment]
Still planning to do this, one day (and if not new then would still be great if someone did it!)

**Comment by pditommaso** [comment]
Was reading this again, I think the real point is having a flag (read env var) that allows enabling/disabling the pull of singularity containers. Agree on that?

**Comment by ewels** [comment]
I'd prefer it as an additional location: check if the container I need is in the read-only location, fetch it to the write-location if not.. So then needs to be a little more than just a flag.

**Comment by pditommaso** [comment]
Therefore I may not fully understand your use. You are planning to have two location for singularity images: 
1. `NXF_SINGULARITY_LIBRARY`: NF checks this path , if the image exists take, otherwise point 2: 
2. `NXF_SINGULARITY_CACHEDIR`: NF check if the image exists, yes => run it, no => pull, store in cachedir, run it. 

Make sense?

**Comment by ewels** [comment]
Yes, exactly 👍🏻 

So `NXF_SINGULARITY_CACHEDIR` is unchanged from its current behaviour. `NXF_SINGULARITY_LIBRARY` is a new additional read-only second location that is checked first, as you say.

**Comment by ewels** [comment]
ps. Whilst we're here, could even do something similar for `$NXF_ASSETS` and have a `$NXF_ASSETS_LIBRARY`. Though in our use case we would like to include the pipeline tag in the directory structure so that multiple versions can coexist, which doesn't currently work with the default assets file path structure.

**Comment by stevekm** [comment]
hi @ewels @pditommaso  I see this is marked as completed and also in the associated ticket https://github.com/nextflow-io/nextflow/issues/1706

However it sounds as if this is only implemented currently via an environment variable `NXF_SINGULARITY_LIBRARYDIR` ; is that the case?

This is really problematic for us because its far more difficult (and far less ideal) to police users' environment variables inside the HPC to try to make sure their Nextflow runs are configured correctly to run on our systems. It would be much preferred if this same configuration was possible to be set via a `nextflow.config` setting. That would allow us to simply provide our users with a supplemental `nextflow.config` file which they could use with `nextflow -c our-settings.config run main.nf ...` in order to easily use our systems. Currently I do not see any mention of `NXF_SINGULARITY_LIBRARYDIR` in the configs for Signularity here;

https://nextflow.io/docs/latest/reference/config.html#singularity

is this the case or would we need a separate Issue to request graduating this to a config option instead of just an env var?

**Comment by pontus** [comment]
This seems to be a documentation bug (which is odd, I believe I've seen it there earlier). `singularity.libraryDir` can be used instead of the environment variable. E.g. https://nextflow.io/docs/stable/container.html also mentions it.

**Comment by ewels** [comment]
Yup, I'm also pretty sure that the config option exists. @christopher-hakkaart would you mind taking a look at this if you get a second please?

Having some kind of automated checks (or auto-generated docs?) to make sure that every config option is documented would be a great thing to look into once we have a new config parser @bentsherman

**Comment by christopher-hakkaart** [comment]
Thanks all; I'll add it to my sprint starting next week. If it was available earlier it _might_ have been accidentally dropped when we did the docs restructure. Either way, it should be quick to add it back in. I'll update here with the PR once it's ready.
